controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,198 @@
+        Preconditions.checkNotNull(gcProcessingSupplier);
+        Preconditions.checkArgument(gcPeriod != null && !gcPeriod.isNegative());
+
+        this.gcName = gcName;
+        this.guardPath = String.format(GUARD_PATH, gcName);
+        this.watch = new AtomicReference<>();
+        this.zkStoreHelper = zkStoreHelper;
+        this.gcProcessingSupplier = gcProcessingSupplier;
+        this.periodInMillis = gcPeriod.toMillis();
+        this.gcExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.gcLoop = new AtomicReference<>();
+        this.currentBatch = new AtomicInteger(0);
+        this.latestVersion = new AtomicInteger(0);
+    }
+
+    @Override
+    protected void doStart() {
+        // Create znode for storing latest batch id. If the batch id exists, get the value from the store.
+        // We will later register watch on the path and keep receiving any changes to its value.
+        RetryHelper.withRetriesAsync(() -> zkStoreHelper.createZNodeIfNotExist(guardPath)
+                        .thenCompose(v -> fetchVersion().thenAccept(r -> currentBatch.set(latestVersion.get())))
+                        .thenAccept(v -> watch.compareAndSet(null, registerWatch(guardPath))),
+                RetryHelper.RETRYABLE_PREDICATE, 5, gcExecutor)
+                .whenComplete((r, e) -> {
+                    if (e == null) {
+                        notifyStarted();
+                        gcLoop.set(Futures.loop(this::isRunning, () -> Futures.delayedFuture(this::process, periodInMillis, gcExecutor), gcExecutor));
+                    } else {
+                        notifyFailed(e);
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            CompletableFuture<Void> gcLoopFuture = gcLoop.updateAndGet(x -> {
+                if (x != null) {
+                    x.cancel(true);
+                    x.whenComplete((r, e) -> {
+                        if (e != null && !(Exceptions.unwrap(e) instanceof CancellationException)) {
+                            log.error("Exception while trying to stop GC {}", gcName, e);
+                            notifyFailed(e);
+                        } else {
+                            notifyStopped();
+                        }
+                    });
+                }
+                return x;
+            });
+
+            if (gcLoopFuture == null) {
+                notifyStopped();
+            }
+        });
+    }
+
+    int getLatestBatch() {
+        return currentBatch.get();
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> process() {
+        return zkStoreHelper.setData(guardPath, new Data<>(new byte[0], latestVersion.get()))
+                .thenComposeAsync(r -> {
+                    // If we reach here, we were able to update the guard and have loose exclusive rights on batch update.
```

Comment:
This could be a method doc.

Reply:
Did not add it as javadoc as this is a private method and only made visible at package level for unit testing. 
Also, the comments are relevant only after we are able to update the guard. So this is more contextual comment as opposed to java doc.